### PR TITLE
Add option to download only via wifi

### LINKED ISSUE #408: Add option to download only on wifi
First of all, thank you for this awesome package.

I want the user to be able to set an option so that the download doesn't happen in the background using the mobile data.

Scenario:
1. User without an internet connection / while using mobile data opens the app.
2. I want to enqueue the download file, but I don't want it to start downloading (if the user is on mobile data).
3. Now the user is done using the app and force closes the app, and after some time user gets connected to the wifi.
4. Only then I want download to happen,  (currently, it start downloading even on the mobile data, as expected)

So, can we have an additional parameter for `FlutterDownloader.enqueue()` for e.g `bool` field `downloadOnlyOnWifi`.

If there is another workaround for this that would be great as well.

Many thanks.

### PR DESCRIPTION
**Pull request by lyio**
+ Add `allowCellular` flag to the `enqueue` that defaults to `true`
+ Add separate `NSURLSession` on iOS that 
sets `allowsCellularAccess` false in its configuration
+ Schedule download requests with one of the 
two sessions depending on the flag of the download
+ Add the flag to the Android plugin and set the 
network constraints to either `NetworkType.CONNECTED` or `NetworkType.UNMETERED` depending on the value of the flag
+ Migrate sql database on Android to keep track of 
the flag

Closes #408

### CODE REVIEW

**Comment by hnvn** [comment]
Thanks @lyio and @kuhnroyal for this PR. Android implementation is good for me but it seems that iOS implementation is missing configuration for DB scheme. On iOS, we have a pre-defined DB scheme in file `download_tasks.sql` in `ios/Assets/` folder

**Comment by lyio** [comment]
@hnvn it works fine on iOS without updating the schema for some reason. I'll see to it that the file is updated, though.

**Comment by lyio** [comment]
@hnvn I have a couple of questions regarding the scheme. I have updated the binary database file to include a new column `allow_cellular`. 

I have not seen anything in the Objective-C code suggesting that any form of migration strategy is applied. On Android the database is cleared and recreated when changes are made, but on iOS I don't see anything like that.

Am I missing something here?

Do we need to add this first? Because otherwise my changes will have no effect on existing installations as the file is only copied over if it does not exist yet.

**Comment by hnvn** [comment]
The iOS implementation comes without migration solution. I currently don't find down any solution for it. That's why I am quit hesitate to add new field to DB schema. New field on DB need to release as a major update

**Comment by lyio** [comment]
The Android solution to the problem seems to be, to simply delete the DB and create it again when the version changes.

**Comment by lyio** [comment]
@hnvn do you have a migration path for iOS in mind?

File: ios/Classes/FlutterDownloaderPlugin.m [hunk 1]
```diff
@@ -624,79 +657,79 @@ - (void) unqueueStatusEvents {
...
     NSString *savedDir = call.arguments[KEY_SAVED_DIR];
     NSString *shortSavedDir = [self shortenSavedDirPath:savedDir];
     NSString *fileName = call.arguments[KEY_FILE_NAME];
     NSString *headers = call.arguments[KEY_HEADERS];
     NSNumber *showNotification = call.arguments[KEY_SHOW_NOTIFICATION];
     NSNumber *openFileFromNotification = call.arguments[KEY_OPEN_FILE_FROM_NOTIFICATION];
+    bool allowCellular = [call.arguments[KEY_ALLOW_CELLULAR] boolValue];
 
-    NSURLSessionDownloadTask *task = [self downloadTaskWithURL:[NSURL URLWithString:urlString] fileName:fileName andSavedDir:savedDir andHeaders:headers];
+    NSURLSessionDownloadTask *task = [self downloadTaskWithURL:[NSURL URLWithString:urlString] fileName:fileName andSavedDir:savedDir andHeaders:headers allowCellular:allowCellular];
```
Comment (lyio) [inline ios/Classes/FlutterDownloaderPlugin.m:670]: After integrating the changes to how the sessions are initialized (in a `if(_isolate)` condition), this always returns `nil`. The plugin is initialized twice, once in the foreground where no sessions are configured and once in the background where there are.

The `enqueue` call is made on the foreground instance though, where the sessions are missing.

I'm having trouble understanding why this is working in the example and on the `master` branch.

Comment (lyio) [reply ios/Classes/FlutterDownloaderPlugin.m:670]: Disregard this, I have found the problem, I think

**Comment by lyio** [comment]
Due to all the changes in the meantime (especially the migration to Kotlin), I have a hard time rebasing this and still getting it to work.

I will close this PR and make a new one.

**Pull request closed without merging**